### PR TITLE
Add endpoint to retrieve order by basket id.

### PR DESCRIPTION
**Pull request by jimabramson**
XCOM-218

@clintonb @rlucioni pls review - it was actually possible to get almost all the needed functionality by trivially subclassing existing stuff (assuming i haven't missed anything).

### CODE REVIEW

File: ecommerce/extensions/api/urls.py [hunk 1]
```diff
@@ -22,5 +23,10 @@
 
 urlpatterns = patterns(
     '',
-    url(r'^orders/', include(ORDER_URLS, namespace='orders'))
+    url(r'^orders/', include(ORDER_URLS, namespace='orders')),
+    url(
+        r'^baskets/{basket_id}/order$'.format(basket_id=BASKET_ID_PATTERN),
```
Comment (jimabramson) [inline ecommerce/extensions/api/urls.py]: deliberately omitted trailing slash; i believe that is supposed to imply a collection of resources rather than a single one.

File: ecommerce/extensions/api/views.py [hunk 2]
```diff
@@ -91,6 +91,15 @@ def get_object(self):
             raise Http404
 
 
+class RetrieveOrderByBasketView(RetrieveOrderView):
```
Comment (clintonb) [inline ecommerce/extensions/api/views.py:94]: Nice! Very clean.

File: ecommerce/extensions/api/urls.py (hunk 1 above)
Comment (clintonb) [inline ecommerce/extensions/api/urls.py]: The trailing slash is mostly cosmetic. Django supports  automatically redirecting paths without the trailing slash to the path with the trailing slash, if you have the slash in `urls.py`. See https://docs.djangoproject.com/en/1.8/ref/settings/#append-slash.

tl;dr You do want the trailing slash.

**Comment by clintonb** [comment]
:+1:  (See comment on trailing slash.)

File: ecommerce/extensions/api/urls.py (hunk 1 above)
Comment (rlucioni) [inline ecommerce/extensions/api/urls.py]: It's worth noting that we have this feature disabled in our base settings. The reasoning was that the redirect from the path without the trailing slash to the path with the trailing slash causes any data submitted in a POST request to be lost.

File: ecommerce/extensions/api/views.py [hunk 3]
```diff
@@ -91,6 +91,15 @@ def get_object(self):
...
+    """ Allow the viewing of Paid Orders by Basket.
```
Comment (rlucioni) [inline ecommerce/extensions/api/views.py]: Minor: Orders are paid by definition, so this should really just read something like "Allow the viewing of Orders by Basket."

File: ecommerce/extensions/api/tests/test_views.py [hunk 4]
```diff
@@ -67,7 +71,10 @@ def test_get_order(self, order_status):
         response = self.client.get(self.url, HTTP_AUTHORIZATION=self.token)
         self.assertEqual(response.status_code, status.HTTP_200_OK)
         result_order = json.loads(response.content)
-        self.assertEqual(OrderSerializer(self.order).data, result_order)
+        # the next step recursively dict-ifies the OrderSerializer's output
+        # otherwise, the raw result of `.data` does not compare properly with `result_order`
+        expected_data = json.loads(JSONRenderer().render(OrderSerializer(self.order).data))
```
Comment (rlucioni) [inline ecommerce/extensions/api/tests/test_views.py]: A slightly more clear approach is to pass the response data to the `OrderSerializer`:

``` python
response_serializer = OrderSerializer(data=response.data)
expected_serializer = OrderSerializer(self.order)

self.assertEqual(response_serializer.data, expected_serializer.data)
```

File: ecommerce/extensions/api/urls.py (hunk 1 above)
Comment (jimabramson) [inline ecommerce/extensions/api/urls.py]: I find the automatic redirection behavior pernicious.  While I personally lean toward omitting trailing slashes when they don't connote a collection, I'm more concerned that things be consistent, and it looks like the convention is to always use the trailing slash, so I'm going to change to doing that here.

File: ecommerce/extensions/api/views.py (hunk 3 above)
Comment (jimabramson) [inline ecommerce/extensions/api/views.py]: Thanks, will tweak.

File: ecommerce/extensions/api/tests/test_views.py (hunk 4 above)
Comment (jimabramson) [inline ecommerce/extensions/api/tests/test_views.py]: After (much) experimenting I've discovered that this also works and seems preferable:

```
response = self.client.get(self.url, HTTP_AUTHORIZATION=self.token)
self.assertEqual(response.data, OrderSerializer(self.order).data)
```

Apparently, the response objects from the test client are more clever than they appear.

Comment (rlucioni) [inline ecommerce/extensions/api/tests/test_views.py]: Looks good, I like it.

**Comment by rlucioni** [comment]
:+1:
